### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive for generic type

### ISSUE THREAD

**Issue by KleaTech** [typescript-eslint/typescript-eslint#6902]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.3&sourceType=module&code=GYVwdgxgLglg9mABAUwB4EMC2AHANsgHgBUA+ACgChFF0AnAcwC5EiAaKxTdbZsmKZJmZEAlIgC8JGmACeExP0ESpizIgD86xAHJtFMQG8AvhTRY8yMmBC5cIgNxA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRQA7iKaRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSJsS6ACYAFAIaYN8dHizCwAMw2plkmlFkKlqgDKxopwybPopAXxAegA

### Repro Code

```typescript
function example<T>(
    arg: T,
    map: (item: T) => any = item => item ?? ''
) {}
example(null);
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No eslint errors.

### Actual Result

`Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined`

### Additional Info

Left-hand side of `??` operator can clearly be null as shown with the `example(null)` call.
Type hint for the `example(null)` call is the following:
`function example<null>(arg: null, map?: (item: null) => any): void`
